mm2src/coins/utxo/utxo_common.rs

```diff
@@ -2374,6 +2416,144 @@ pub fn send_raw_tx_bytes(
+                height: None,
+            });
+        }
+    }
+
+    let txids_not_passed: HashSet<H256Json> = tx
+        .inputs()
+        .iter()
+        .filter(|input| !unspents.iter().any(|u| u.outpoint.hash == input.previous_output.hash))
+        .map(|input| input.previous_output.hash.into())
+        .collect::<HashSet<_>>();
+
+    // If some previous utxos are not provided in the params load them from the chain
+    if !txids_not_passed.is_empty() {
+        let txids_reversed = txids_not_passed
+            .iter()
+            .map(|txid| txid.reversed())
+            .collect::<HashSet<_>>(); // reverse hashes to send to electrum
+        let prev_txns_loaded = utxo_common::get_verbose_transactions_from_cache_or_rpc(coin.as_ref(), txids_reversed)
+            .await
+            .map_err(|err| RawTransactionError::InvalidParam(err.get_inner().to_string()))?;
+
+        let mut unspents_loaded = vec![];
+        for input in tx.inputs() {
+            if let Some(prev_tx) = prev_txns_loaded
+                .iter()
+                .find(|p| (*p.0).reversed() == input.previous_output.hash.into())
+            {
+                let prev_tx = prev_tx.1.to_inner();
+                if (input.previous_output.index as usize) < prev_tx.vout.len() {
+                    if prev_script.is_none() {
+                        prev_script = Some(
+                            prev_tx.vout[input.previous_output.index as usize]
+                                .clone()
+                                .script
+                                .hex
+                                .into(),
+                        ); // get first prev script assuming all are the same
+                    }
+                    let prev_amount = prev_tx.vout[input.previous_output.index as usize]
+                        .value
+                        .ok_or_else(|| {
+                            RawTransactionError::NonExistentPrevOutputError(String::from("no previous amount"))
+                        })?;
+                    let prev_amount = prev_amount.try_into().expect("Amount conversion must succeed");
+
+                    unspents_loaded.push(UnspentInfo {
+                        outpoint: OutPoint {
+                            hash: input.previous_output.hash,
+                            index: input.previous_output.index,
+                        },
+                        value: sat_from_big_decimal(&prev_amount, coin.as_ref().decimals)
+                            .expect("Conversion to satoshi from bigdecimal must be valid"),
+                        height: None,
+                    });
+                    continue;
+                }
+            }
+            return Err(RawTransactionError::NonExistentPrevOutputError(format!(
+                "{}/{}",
+                input.previous_output.hash, input.previous_output.index
+            ))
+            .into());
+        }
+        unspents.extend(unspents_loaded.into_iter());
+    }
```

Comment:
This can be made a separate function to make the code more readable, e.g.
```rust
async fn load_missing_inputs_from_cache_or_chain<T: AsRef<UtxoCoinFields>>(
    coin: &T,
    tx: &UtxoTx,
    unspents: &mut Vec<UnspentInfo>,
) -> Result<(), RawTransactionError>  {
    let txids_not_passed: HashSet<H256Json> = tx
        .inputs()
        .iter()
        .filter(|input| !unspents.iter().any(|u| u.outpoint.hash == input.previous_output.hash))
        .map(|input| input.previous_output.hash.into())
        .collect::<HashSet<_>>();

    if txids_not_passed.is_empty() {
        return Ok(());
    }

    // Rest of the code
}
```

Please also look for other parts of this function that can be refactored for better readability/maintainability.

Reply:
It's very hard to refactor more, I tried a few things myself that are not ideal. I think it's not worth it to spend more time trying to refactor this function.